### PR TITLE
Arborist to template

### PR DESCRIPTION
**Pull request by Spayralbe**
Adds functionality for transforming the JSON tree from an Arborist tree file/tree URL to an Excel tree structure template. The code can be executed as a standalone script or from tmtk.toolbox.

### CODE REVIEW

File: tmtk/toolbox/arborist_to_template.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+def _get_study_name_from_url(tree_input_url):
+    # Name of study is the URL part after "trees"
+    parts = tree_input_url.split('/')
+    trees_index = parts.index('trees')
+    study_name = parts[trees_index+1]
+    return study_name
+
+
+def _get_json_tree_from_file(tree_input):
+    check = click.Path(exists=True, dir_okay=False, readable=True)
```
Comment (wibo-pipping) [inline tmtk/toolbox/arborist_to_template.py]: Catch exception here and return nice error message. click will probably produce a stack trace where you only need the information that the specified path to a json file does not exist.

File: tmtk/toolbox/arborist_to_template.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        self.df = pd.DataFrame()
+
+    def transform_to_template(self):
+        self.add_root_to_json_tree()
+        self.loop_through_nodes(self.raw_tree)
+        self.add_missing_metadata_columns()
+        self.order_columns()
+        self.add_mandatory_columns()
+
+        writer = pd.ExcelWriter(self.output_path)
```
Comment (wibo-pipping) [inline tmtk/toolbox/arborist_to_template.py]: Split the saving/writing to excel to a separate function. This is probably something you want to do after transforming and gives you the option to also inspect the template that is going to be saved before actual saving it.

Optional could include the function to save it at the and of the transform_to_template function but by moving it to a separate function provides more flexibility to a power user

File: tmtk/toolbox/arborist_to_template.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        self.df.insert(loc=0, column='Column number', value=np.nan)
+        self.df.insert(loc=0, column='Sheet name/File name', value=np.nan)
+        self.df.insert(loc=0, column='tranSMART data type', value='Low-dimensional')
+
+    def order_columns(self):
+        level_order = ['Level {}', 'Level {} metadata tag', 'Level {} metadata value']
+        max_level = max(
+            [int(col.split()[1]) for col in self.df.columns if col.startswith('Level') and not 'metadata' in col])
+        ordered_cols = [col.format(i) for i in range(1, max_level + 1) for col in level_order]
+        self.df = self.df.reindex(ordered_cols, axis=1)
```
Comment (wibo-pipping) [inline tmtk/toolbox/arborist_to_template.py]: This is only replacing the name of the columns right? are the column in the df always in the correct order? If not sure have to check for this.

File: tmtk/toolbox/arborist_to_template.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    def add_metadata_to_node_level(self, node):
+        """Add metadata to actual node instead of it being a child node"""
+        if any([child_node['text'] == 'Tags' for child_node in node['children']]):
+            node['metadata'] = [child['data']['tags'] for child in node['children'] if child['text'] == 'Tags'][0]
+            # Sort tags by weight
+            node['metadata'] = sorted(node['metadata'].items(), key=lambda kv: int(kv[1][1]))
+
+    def loop_through_nodes(self, node, depth=1):
+        if not self.skip_metadata:
```
Comment (wibo-pipping) [inline tmtk/toolbox/arborist_to_template.py]: please add a docstring to this one.

File: tmtk/toolbox/arborist_to_template.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            row_index = 0
+        # If all >= levels on this row are empty, add node on the same row
+        elif all(self.df.loc[last_row_index, cols_equal_or_higher_level].isnull()):
+            row_index = last_regular_node_row_index
+        # Otherwise insert node on a new row
+        else:
+            row_index = last_row_index + 1
+        return row_index
+
+    def insert_node_data_into_df(self, node, depth):
```
Comment (wibo-pipping) [inline tmtk/toolbox/arborist_to_template.py]: Nice to have - Doc string or comments in the function.
Shortly describe what regular_node_cols, last_regular_node_row_index and cols_equal_or_higher_level are being used for. Took me some time to understand what is going on.

File: tmtk/toolbox/arborist_to_template.py (hunk 1 above)
Comment (Spayralbe) [reply tmtk/toolbox/arborist_to_template.py]: @wibo-pipping Click handles this quite nicely, so you actually don't get a stack trace when any of these checks fail. Of course there are many other things that can be wrong with the provided input and then you will get a stack trace.

File: tmtk/toolbox/arborist_to_template.py (hunk 3 above)
Comment (Spayralbe) [reply tmtk/toolbox/arborist_to_template.py]: The names stay the same, but they are put in the right order (i.e. starting with level 1, then level 1 metadata tag etc.).

**Comment by jochemb** [comment]
Closing for now, will be reopened after fixes to branch

**Pull request merged**